### PR TITLE
[WIP] refactor sync/analyze stage 1a

### PR DESCRIPTION
**Pull request by thearthur**
first pass, this is incomplete because the new analysis is not plumbed
into all the places where we actually need to run analyze.

### CODE REVIEW

File: src/metabase/cache_database.clj [hunk 1]
```diff
@@ -0,0 +1,76 @@
+(ns metabase.cache-database
```
Comment (salsakran) [inline src/metabase/cache_database.clj]: this should be either "cache-values" or "field-values" or "cached-field-values". any opinions @camsaul ?

Comment (salsakran) [inline src/metabase/cache_database.clj]: or maybe "cache-field-values"

**Review by salsakran — COMMENTED** [review]
couple things jump out at me

1. the naming of the "cache" section is confusing. We're *not* caching databases, tables or fields here. We're just extracting a set of field values for a subset of fields in each of the above. We should be explicit about this.
2. It seems like this re-fetches field values for both analyze and field-value-caching. we should figure out a way to make it possible to chain these together on the first run of things. 
eg. the first time I run analysis on a freshly added db 
I'm going to do both a sync (that doesn't need to touch field values), and then a fingerprint and a field-values-cache (which do). The latter two should be chainable off a single fetch of field values.

File: src/metabase/cache_database.clj (hunk 1 above)
Comment (camsaul) [inline src/metabase/cache_database.clj]: yeah. it shouldn't be a top-level namespace either IMO. It should be `metabase.sync.something`. That way it's clear it's part of the sync part of the app.

@thearthur think about that chart you drew up showing how the different parts of the app interconnect, and how much more confusing that would have been to make if everything was just a top-level namespace

File: test/metabase/sync_database/analyze_test.clj [hunk 2]
```diff
@@ -18,18 +18,6 @@
 
 ;; test:cardinality-and-extract-field-values
 ;; (#2332) check that if field values are long we skip over them
-(expect
-  {:values nil}
```
Comment (salsakran) [inline test/metabase/sync_database/analyze_test.clj]: why are these deleted?

File: src/metabase/sync_database/analyze.clj [hunk 3]
```diff
@@ -157,86 +53,42 @@
...
+  (let [values (db/select 'FieldValues :id (:id table))
+        #_values-to-test-for-json-and-email #_(take driver/max-sync-lazy-seq-results 
+                                                    (driver/field-values-lazy-seq driver field))]
+      {:id                                (:id field) ;; check if this should be field or table id
+       :field-percent-urls      (u/try-apply (:field-percent-urls driver) field)
+       :field-percent-json      (if (values-are-valid-json? values) 100 0)
+       :field-percent-email     (if (values-are-valid-emails? values) 100 0)
+       :field-avg-length        (u/try-apply (:field-avg-length driver) field)
+       :visibility_type         (:visibility_type field)
+       :values                  values
```
Comment (salsakran) [inline src/metabase/sync_database/analyze.clj]: Values should not be in the fingerprint. They will be both large and potentially sensitive.

Comment (thearthur) [inline src/metabase/sync_database/analyze.clj]: left that in while debugging the sync tests  (which are still broken)

File: test/metabase/sync_database/analyze_test.clj (hunk 2 above)
Comment (thearthur) [inline test/metabase/sync_database/analyze_test.clj]: moved to metabase.sync-database.cached-values-test

File: src/metabase/sync_database/analyze.clj [hunk 4]
```diff
@@ -157,86 +53,53 @@
...
+                           (recur (if valid? (inc valid-count) valid-count)
+                                  (inc non-nil-count)
+                                  more)))))
+
+(defn field-fingerprint [driver table field]
+  (let [values (take driver/max-sync-lazy-seq-results
+                     (driver/field-values-lazy-seq driver field))]
+    {:id                      (:id field) ;; check if this should be field or table id
+     :field-percent-urls      (percent-valid-urls values)
+     :field-percent-json      (if (values-are-valid-json? values) 100 0)
```
Comment (salsakran) [inline src/metabase/sync_database/analyze.clj]: this should return the percent, not turn it into a binary feature.

**Comment by thearthur** [comment]
The new approach to calculating the field "fingerprint" is to remove all the driver specific logic and derive all the field fingerprint values from the first 1000 entries for the field in the table.

File: src/metabase/driver.clj [hunk 5]
```diff
@@ -20,7 +20,7 @@
   "The maximum number of values we should return when using `field-values-lazy-seq`.
    This many is probably fine for inferring special types and what-not; we don't want
    to scan millions of values at any rate."
-  10000)
+  1000)
```
Comment (salsakran) [inline src/metabase/driver.clj]: we should experiment with this.

**Comment by thearthur** [comment]
This is not the most current PR for this issue. new conversation should move to: https://github.com/metabase/metabase/pull/5363
